src/action.c

```diff
@@ -743,6 +826,12 @@ actions_run(struct view *activator, struct server *server,
 				desktop_focus_output(output_from_name(server, output_name));
 			}
 			break;
+		case ACTION_TYPE_FOR_EACH:
+			struct view *v, *next;
+			wl_list_for_each_safe(v, next, &server->views, link) {
```

Comment:
Iterating over all views for the “else” actions is likely to cause problems with some xwayland windows.

Suggest using the new “desktop_views_append” API (PR close to merge.

Or at least check that view has surface and has been mapped at some point.

Reply:
I'll check the desktop_views_append() API, thanks.

Reply:
Seems like desktop_views_append() only works for surfaces on the current desktop, while ForEach should work on all desktops. Checking for mapped surfaces maybe an option. Could you elaborate on how to do it and what happens if I don't?

Reply:
You can use:

```
struct view *view;
wl_list_for_each(view, &server->views, link)
{
	if (!isfocusable(view)) {
		continue;
	}
	// ...then do whatever with view
}
```

If you don't call `isfocusable()` you run the risk of segfault because some xwayland clients create views which are never ever mapped and can set view->surface to NULL at any time.

The new API (https://github.com/labwc/labwc/pull/1024/files) is not just for current workspace. It can be used like this:

```
struct wl_array views;
wl_array_init(&views);
desktop_views_append(server, &views, 0 /* no criteria */);

struct view **view;
wl_array_for_each(view, &views) {
	// Do something with *view
}
wl_array_release(&views);
```

Reply:
Okay, the `desktop_` part completely threw me off.